File: medium/45. Jump Game II/45. Jump Game II.ts
function jump(nums: number[]): number {
    let jumps = 0
    let reachable = 0
    let end = 0

    for (let i = 0; i < nums.length - 1; i++) {
        reachable = Math.max(reachable, nums[i] + i)

        if (end === i) {
            end = reachable
            jumps += 1
        }

        if (end >= nums.length - 1)
            break
    }

    return jumps
};

File: medium/45. Jump Game II/45. Jump Game II.py
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        reachable = 0
        current_end = 0
        steps = 0

        for i, num in enumerate(nums[:-1]):
            reachable = max(reachable, num + i)

            if current_end == i:
                current_end = reachable
                steps += 1

            if current_end >= len(nums) - 1:
                break

        return steps


solution = Solution()
print(solution.jump([2, 3, 1, 1, 4]))
print(solution.jump([2, 3, 0, 1, 4]))
print(solution.jump([1]))
print(solution.jump([1, 2, 1, 1, 1]))
print(solution.jump([2, 1]))
